text/0000-block-manager.md

```diff
@@ -0,0 +1,367 @@
+blocks to these stores.
+
+Persisting blocks is handled by the following methods:
+
+`add_store(store_name: String, store: Store)`
+
+    Take ownership of the given store and enable it for persisting blocks. The
+    store must be referenced from other methods with `store_name`.
+
+    Raises an exception if a store with `store_name` already exists.
+
+`persist(head: String, store_name: String)`
+
+    Atomically ensure that all blocks on the branch starting with head are
+    in the store.
+
+## Block Store Interface
+
+The __block store__ interface is the interface that must be implemented by an
+object in order for it to be added to the block tree as a "store". The
+following is a listing of the interface's methods:
+
+`put(blocks: List<Block>)`
+
+    Atomically add the given blocks to the store.
+
+`get(block_ids: List<String>) -> Iter<Block>`
+
+    Atomically get the blocks with the given ids.
+
+`iter() -> Iter<Block>`
+
+    Create an iterator over all blocks in the store.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The following provides more detail on how the API present above is implemented.
+
+## Internal Representation
+
+Internally, the block manager consists of:
+
+- A __main cache__ which, at a minimum, contains all blocks that have not been
+  persisted.
+- A map of store names to stores.
+
+Internally, the commit store consists of two components:
+
+- A database object which is responsible for persisting the committed fork to
+  disk and retrieving committed blocks from disk.
+- An __LRU cache__, which is responsible for ensuring that frequently accessed
+  blocks do not incur database reads. All writes are done directly to disk, but
+  this behavior could be changed in the future. The LRU cache keeps the last N
+  recently used blocks, where "used" can mean either a read or a write.
+
+## Ensuring Integrity
+
+As is heavily implied by the API, the block manager uses __reference counting__
+internally to guarantee integrity and to allow external components to place
+holds on specific blocks. For every block managed by the block manager, a
+reference count is maintained (sometimes implicitly).
+
+This one reference count is a sum of both:
+
+1. The count of all other blocks in the block manager that depend directly
+   on this block
+2. The count of all calls to `ref_block` minus the count of all calls to
+   `unref_block`
+
+In order to avoid keeping a reference count and copy of every persisted block
+in the main cache, __anchors__ are placed in the main cache to indicate that
+the rest of the branch has been persisted to some store. An __anchor__ is used
+to represent a block that is in a store and not in the __main cache__ and
+consists of:
+
+- The block id of the block that is in the store
+- A list of names of stores that the block is in
+- A reference count of blocks that depend on the anchor but are not in a store
+  the block is in
```

Comment:
This seems confusing to have reference counting on the anchor and on the block. Would it make sense to just have reference counting managed by the tree? I'm assuming all blocks are referenced in the tree(s)?